src/main/scala/com/fasterxml/jackson/module/scala/modifiers/ScalaTypeModifier.scala

```diff
@@ -0,0 +1,33 @@
+package com.fasterxml.jackson.module.scala.modifiers
+
+import java.lang.reflect.{ParameterizedType, Type}
+
+import com.fasterxml.jackson.databind.JavaType
+import com.fasterxml.jackson.databind.`type`._
+import com.fasterxml.jackson.module.scala.JacksonModule
+
+class ScalaTypeModifier extends TypeModifier {
+  override def modifyType(javaType: JavaType,
+                          jdkType: Type,
+                          context: TypeBindings,
+                          typeFactory: TypeFactory): JavaType = {
+
+
+    if (javaType.isTypeOrSubTypeOf(classOf[collection.Map[_, _]])) {
+      MapLikeType.upgradeFrom(javaType, javaType.containedTypeOrUnknown(0), javaType.containedTypeOrUnknown(1))
+    } else if (javaType.isTypeOrSubTypeOf(classOf[collection.TraversableOnce[_]])) {
+      CollectionLikeType.upgradeFrom(javaType, javaType.containedTypeOrUnknown(0))
+    } else if (javaType.isTypeOrSubTypeOf(classOf[Option[_]])) {
+      ReferenceType.upgradeFrom(javaType, javaType.containedTypeOrUnknown(0))
```

Comment:
So, let me restate what you're saying, to see if I understand correctly:

Scala has three classes in its implementation `Option` (as opposed to Java 8's singular class `Optional`): `Option`, which is an abstract base class, `Some`, which is a class that represents the presence of a value, and `None`, which indicates the absence. When someone attempts to serialize a value of `Some("Foo")` (which has type `Some[String]`), Jackson will recognize that `Some[String]` derives from `Option[String]`, and resolve the class `Option.class` first.

When the Type Modifier modifies `Option` to be a reference type, then Jackson will automatically consider `Some` to be a reference type as well, thus making a check for subtyping unneeded, though the type modifier could probably use an initial check to see if the type is already considered a container or reference type.

Is this the gist?